doc/neps/nep-0029-dtype-as-type.rst

```diff
@@ -0,0 +1,303 @@
+become an abstract base class, and a number of new classes will be created with
+a hierarchy like scalars. They will support subclassing. A future NEP may
+propose unifying the scalar and dtype type systems, but that is not a goal of
+this NEP.  The changed dtype will:
+
+- facilitate extending dtypes, typically for things like categoricals, novel
+  representations like datetime or IP addresses, or adding attributes like
+  units directly on the dtypes.
+- Allow replacement of the switch-type code we have in places like arrayprint_,
+  PyArray_AdaptFlexibleDtype_, and PyArray_CanCastSafely_ (just a few of many),
+  with method-based overrides, simplifying internal logic and allowing
+  subclassed dtypes to override the functionality.
+
+.. _arrayprint: https://github.com/numpy/numpy/blob/v1.16.0rc1/numpy/core/arrayprint.py#L418
+.. _PyArray_AdaptFlexibleDtype: https://github.com/numpy/numpy/blob/v1.16.0rc1/numpy/core/src/multiarray/convert_datatype.c#L164 
+.. _PyArray_CanCastSafely: https://github.com/numpy/numpy/blob/v1.16.0rc1/numpy/core/src/multiarray/convert_datatype.c#L398
+
+Overall Design
+--------------
+
+The ``Dtype`` class (and any subclass without ``itemsize``) is effectively an
+abstract base class, as it cannot be used to create instances. A class
+hierarchy is used to add datatype-specific functionality such as ``names`` and
+``fields`` for structured dtypes. The current behaviours are preserved:
+
+- ``np.dtype(obj, align=False, copy=False)`` calls ``arraydescr_new`` with
+  various types of ``obj``:
+  - ``int``, ``np.genenric`` scalar classes, list or dict are parsed into
+    appropriate dtypes
+- singletons are returned where appropriate
+
+Additionally, dtype subclasses are passed through to the subclass ``__new__``
+
+A prototype without error checking, without options handling, and describing
+only ``np.dtype(np.uint8)`` together with an overridable ``get_format_function``
+for ``arrayprint`` looks like::
+
+    import numpy as np
+
+    class Dtype():
+        def __new__(cls, *args, **kwargs):
+            if len(args) == 0: 
+                # Do not allow creating instances of abstract base classes
+                if not hasattr(cls, 'itemsize'):
+                    raise ValueError("cannot create instances of "
+                                     f"abstract class {cls!r}")
+                return super().__new__(cls, *args, **kwargs)
+            # This is reached for Dtype(np.uint8, ...).
+            # Does not yet handle align, copy positional arguments
+            obj = args[0]
+            if isinstance(obj, int):
+                return dtype_int_dict[obj]
+            elif isinstance(obj, type) and issubclass(obj, np.generic):
+                return dtype_scalar_dict[obj]
+            else:
+                # Dtype('int8') or Dtype('S10') or record descr
+                return create_new_descr(cls, *args, **kwargs)
+
+    class GenericDtype(Dtype):
```

Comment:
There seems to be no benefit to having `GenericDtype` any more. I thought the argument before that you don't necessarily want to override what `int` means, but want to be sure that `itemsize` exists, was logical, so I would move some of the `__new__` above here (or stick with the metaclass, it really wasn't bad and could be used to auto-fill methods based on `itemsize`, etc.)